common/js/src/messaging/single-message-based-channel-unittest.js

```diff
@@ -0,0 +1,310 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+goog.require('GoogleSmartCard.SingleMessageBasedChannel');
+goog.require('GoogleSmartCard.MessageChannelPinging.Pinger');
+goog.require('GoogleSmartCard.MessageChannelPinging.PingResponder');
+goog.require('GoogleSmartCard.TypedMessage');
+goog.require('goog.Promise');
+goog.require('goog.async.Delay');
+goog.require('goog.testing.PropertyReplacer');
+goog.require('goog.testing.jsunit');
+goog.require('goog.testing.mockmatchers');
+
+goog.setTestOnly();
+
+goog.scope(function() {
+
+/** @const */
+var GSC = GoogleSmartCard;
+
+/** @const */
+var SingleMessageBasedChannel = GSC.SingleMessageBasedChannel;
+
+/** @const */
+var Pinger = GSC.MessageChannelPinging.Pinger;
+
+/** @const */
+var PingResponder = GSC.MessageChannelPinging.PingResponder;
+
+/** @const */
+var TypedMessage = GSC.TypedMessage;
+
+/** @const */
+var EXTENSION_ID = 'DUMMY_EXTENSION_ID';
+
+/** @const */
+var TEST_MESSAGES = [
+    new TypedMessage('service x', {}),
+    new TypedMessage('service x', {some_data: 1}),
+    new TypedMessage('service y', {})];
+
+/** @const */
+var channelId = PingResponder.generateChannelId();
+
+/** @const */
+var pingResponseMessage = new TypedMessage(
+    PingResponder.SERVICE_NAME,
+    PingResponder.createMessageData(channelId)).makeMessage();
+
+/**
+ * Mock matcher that matches strings.
+ * @const
+ */
+var isStringMockMatcher = new goog.testing.mockmatchers.ArgumentMatcher(
+    function(value) {
+      return typeof value === 'string';
+    }, 'isStringMessage');
+
+/**
+ * Mock matcher that matches only the ping messages.
+ * @const
+ */
+var isPingMessageMockMatcher = new goog.testing.mockmatchers.ArgumentMatcher(
+    function(value) {
+      var typedMessage = TypedMessage.parseTypedMessage(value);
+      return typedMessage && typedMessage.type == Pinger.SERVICE_NAME;
+    }, 'isPingMessage');
+
+var globalChannel;
+
+/**
+ * Setup the mock for chrome.runtime.sendMessage.
+ * @param {!goog.promise.Resolver} testCasePromiseResolver
+ * @return {goog.testing.PropertyReplacer}
+ */
+function setupSendMessageMock(testCasePromiseResolver) {
+  var propertyReplacer = new goog.testing.PropertyReplacer;
+  propertyReplacer.setPath(
+      'chrome.runtime.sendMessage',
+      goog.testing.createFunctionMock('sendMessage'));
+  testCasePromiseResolver.promise.thenAlways(
+      propertyReplacer.reset, propertyReplacer);
+
+  return propertyReplacer;
+}
+
+/**
+ * Setup ping responding to messages going through chrome.runtime.sendMessage
+ * (essentially intercepts the message and sends a response back).
+ */
+function setupSendMessagePingResponding() {
+  // Type cast to prevent Closure compiler from complaining.
+  /** @type {?} */ chrome.runtime.sendMessage;
+
+  chrome.runtime.sendMessage(isStringMockMatcher, isPingMessageMockMatcher)
```

Comment:
It shouldn't be hard to make a bit stricter test by verifying that the passed string is actually equal to EXTENSION_ID.